src/settings/advanced.c

```diff
@@ -53,6 +54,145 @@ iap_advanced_hildon_number_editor_new()
+    GConfClient* client = gconf_client_get_default();
+
+    char* key = g_strdup_printf("/system/osso/connectivity/srv_provider/%s/available_ids", service_type);
+    if (!key)
+        goto out;
+
+    providers = gconf_client_get_list(client, key, GCONF_VALUE_STRING, &error);
+
+    if (error != NULL) {
+        ULOG_WARN("Unable to list available_ids for %s: %s\n", service_type, error->message);
+        g_clear_error(&error);
+
+        goto out;
+    }
+
+out:
+    if (key != NULL)
+        g_free(key);
+    g_object_unref(client);
+
+    return providers;
+}
+
+static void iap_advanced_provider_load_providers(GtkWidget *widget_types, GtkWidget *widget_ids) {
+    GSList* service_types = get_gconf_service_types();
+    GSList *l;
+
+    if (!service_types) {
+        ULOG_WARN("iap_advanced_provider_load_providers: Unable to find any service_types in gconf");
+    }
+
+    gtk_combo_box_append_text(GTK_COMBO_BOX(widget_types), "None");
+
+    for (l = service_types; l; l = l->next) {
+        gtk_combo_box_append_text(GTK_COMBO_BOX(widget_types), (const gchar*)l->data);
+
+        /* We don't need to load service ids here, since we will change them once
+         * the service_type gets set by the mapper, in
+         * iap_advanced_provider_changed_cb, after which the mapper will set it to a
+         * valid value, if any */
+    }
+
+    g_slist_free_full(service_types, g_free);
+    return;
+}
+
+static gboolean
+iap_advanced_provider_changed_cb(GtkWidget *widget, struct iap_wizard_advanced *adv)
+{
+    GtkWidget *service_type_widget = g_hash_table_lookup(adv->widgets, "PROVIDER_SERVICE_TYPE");
+    GtkWidget *service_id_widget = g_hash_table_lookup(adv->widgets, "PROVIDER_SERVICE_ID");
+
+    if (service_type_widget == widget) {
+        GtkTreeIter iter;
+        const char* service_type;
+        GSList *service_ids, *ll;
+        gboolean valid;
+        gint ids = 0;
+
+        GtkTreeModel* model_id = gtk_combo_box_get_model(GTK_COMBO_BOX(service_id_widget));
+        gtk_list_store_clear(GTK_LIST_STORE(model_id));
+
+        GtkTreeModel* model_type = gtk_combo_box_get_model(GTK_COMBO_BOX(service_type_widget));
+        valid = gtk_combo_box_get_active_iter(GTK_COMBO_BOX(service_type_widget), &iter);
+        if (!valid) {
+            ULOG_WARN("Could not get iterator for service_type_widget model");
+            return TRUE;
+        }
+        gtk_tree_model_get(GTK_TREE_MODEL(model_type), &iter, 0, &service_type, -1);
+
+        service_ids = get_gconf_service_ids(service_type);
+
+        for (ll = service_ids; ll; ll = ll->next) {
+            gtk_combo_box_append_text(GTK_COMBO_BOX(service_id_widget), (const char*)ll->data);
+            ids++;
+        }
+
+        // We could improve this by trying to find the current service_id in the
+        // IAP in gconf, provided the service_type currently selected is the
+        // same as the one in the IAP, but that's too much effort.
+        if (ids > 0)
```

Comment:
but this should be done, otherwise you may save the IAP with wrong service_type, no?

Reply:
I don't think so, this only happens when users actively change the service_type and then do not look at the service_id at all any more. One problem I am facing is that I don't actually have access to the IAP id in the code (as far as I could see), so reading this was tricky.

Reply:
In any case it will always be shown to the users, it's just a minor UX inconvenience. I'll see if I can fix it, but it doesn't silently select something wrong.